### DISCUSSION TITLE
Question about the LED

### DISCUSSION THREAD

**Discussion by Micha213** [darthcloud/BlueRetro#22] [Q&A]
Hi.

I wanted to design a small circuit board for the esp32 d1 mini BlueRetro receiver for PS1/2 that I built. The circuit board should make wiring easier, provide a place for an LDO and lead the buttons and LEDs to the outside. 

But I'm not exactly sure about the LED. The power LED is not a problem. The ESP32 d1 mini and also the ESP32 devkit have an LED connected to IO02 and GND.
With the BlueRetro devkit, IO02 goes into the auto programming circuit and instead an LED is connected to IO17 and 3.3V. 
I temporarily connected both LEDs (IO02 + IO17) to test them, but it seems that neither is doing anything. 

Is the LED being used for anything or is something planned? Which is the right one, IO17 or IO02? 
If the LED is not used, is it possible to use it to display the analog mode of the PS1/2 controller? LED on for analog mode and off for digital mode, like the original Dualshock 1/2.

**Comment by darthcloud** [comment]
IO2 is reserved for SD card, the LED is IO17 and is used to display if the adapter is in error. So normaly it should be off, except on briefly at boot or while holding reset button.

For your specific use case since your not using SD I could hook up IO2 to the P1 analog state. I'll see what I can do.

**Comment by darthcloud** [comment]
[v0.15.1_WIP.zip](https://github.com/darthcloud/BlueRetro/files/6320867/v0.15.1_WIP.zip)
Untested but try this and let me know if it work :)

**Comment by Micha213** [comment]
Hi.

Thanks for the very quick implementation. 
I tried it this morning. At first the LED did not light up. I had the external LED connected like the one on the board: IO2 - LED - Resistor - GND. After I turned the LED and connected the other end to + 3.3V instead of GND, it worked:
![ledio2](https://user-images.githubusercontent.com/46410328/115055908-6d640100-9ee2-11eb-9f7b-0302cb461932.png)


The LED always lights up very weakly at the beginning, until I turn the analog mode on or off for the first time. From then on it runs normally and also goes out completely when analog mode is deactivated. Does IO2 not have a defined state when it starts?

**Reply by darthcloud** [reply]
I took this D1 mini schematic here: https://github.com/artofcircuits/Store/blob/main/WeMOS%20D1%20Mini%20ESP32/WEMOS-D1-MINI-ESP32-SCH.jpg

https://raw.githubusercontent.com/artofcircuits/Store/main/WeMOS%20D1%20Mini%20ESP32/WEMOS-D1-MINI-ESP32-SCH.jpg

Its not the board you use? Can you link to the board schematic you use?

I will add init and post new build soon.

**Comment by Micha213** [comment]
Yes, this is the right board. Mine just isn't branded with wemos. One board says "mh-et live" and the other "hw-665". I'm not sure if there are different versions of it. When I was looking for a Shematic, I always found something else.
Tonight, when I'm back home, I'll check the other boards that I still have here, how the LED is connected there. Give me a little more time, I'll check it out. ;)

**Comment by darthcloud** [comment]
[v0.15.1_WIP2.zip](https://github.com/darthcloud/BlueRetro/files/6327352/v0.15.1_WIP2.zip)

This should init the LED and require same circuit as before.

**Comment by Micha213** [comment]
I have another question about the LED on IO17.
I've never seen an LED connected like this. How exactly does it work? Is the LED switched when IO17 output is high (LED off) and input (LED on)?
Can the LED be wired normally without having to change anything in the code?
![led17](https://user-images.githubusercontent.com/46410328/125164963-2d38b580-e195-11eb-9e47-a2373208bf79.png)

**Reply by darthcloud** [reply]
Its wired this way for the LED to be on if the pin is high impedance (input). So if software fail to boot the LED is on indicating an error. First thing software do is to put pin output low to turn off.

**Reply by damontgomery** [reply]
Thanks for this explanation. I found it very helpful.

I'm pretty new to electronics and have only ever seen (in my very limited experience) `Vo (or PIN OUT) -> R -> LED -> GND`. For context, it probably took me ~2h to figure out what size resistor to use this afternoon with a standard red LED. :)

Since the rest of the project seems pretty straight forward. It might be helpful to add this note to the https://github.com/darthcloud/BlueRetro/wiki/BlueRetro-DIY-Build-Instructions page.

I haven't built the system yet (waiting for cables), but thanks for all the work and community support!

...

If it helps anyone my standard red LED had a voltage drop of around 1.8V and it seems to work with a 220ohm resistor. This produced about 12mA of current which is supposed to be about 1/2 of the maximum 20mA allowed for the standard red LED. It's on, but it is not bright. :)
